### PR TITLE
Add ESIM model

### PR DESCRIPTION
**Pull request by Victor0118**
Reference:
paper: Enhanced {LSTM} for Natural Language Inference
code: https://github.com/lanwuwei/SPM_toolkit/tree/master/ESIM

@likicode @daemon Could you take a look at this PR?

### CODE REVIEW

**Comment by Victor0118** [comment]
## ESIM without tree after tuning on SICK dataset

==> esim_sick_1e-4_0.001_0.3_0.2.log <==
INFO -          pearson_r       spearman_r      mse     KL-divergence loss
INFO - test     0.878273        0.823042214423  0.25375571846961975     0.551383518848454

## ESIM-Tree after tuning on SICK dataset
==> esim_tree_sick_1e-4_0.001_0.3_0.2.log <==
INFO -          pearson_r       spearman_r      mse     KL-divergence loss
INFO - test     0.878069        0.820418404008  0.2360289841890335      0.4703939018613406

File: esim/__main__.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    saved_model_evaluator = EvaluatorFactory.get_evaluator(dataset_cls, model, embedding, loader, batch_size, device,
+                                                           keep_results=keep_results)
+    scores, metric_names = saved_model_evaluator.get_scores()
+    logger.info('Evaluation metrics for {}'.format(split_name))
+    logger.info('\t'.join([' '] + metric_names))
+    logger.info('\t'.join([split_name] + list(map(str, scores))))
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='PyTorch implementation of Multi-Perspective CNN')
```
Comment (daemon) [inline esim/__main__.py]: `description` is good to include.

File: esim/__main__.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                        help='use TensorBoard to visualize training (default: false)')
+    parser.add_argument('--run-label', type=str, help='label to describe run')
+    parser.add_argument('--keep-results', action='store_true',
+                        help='store the output score and qrel files into disk for the test set')
+
+    args = parser.parse_args()
+
+    device = torch.device(f'cuda:{args.device}' if torch.cuda.is_available() and args.device >= 0 else 'cpu')
+
+    random.seed(args.seed)
```
Comment (daemon) [inline esim/__main__.py]: Did we fold this part into a common module? E.g., `import common.utils; common.utils.set_all_seeds(x)`?

File: esim/model.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
+import sys
+import math
+import torch
+import numpy as np
+import torch.nn as nn
+from esim.util import *
```
Comment (daemon) [inline esim/model.py]: PEP8 import order.

File: esim/model.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+import torch.nn.functional as F
+from torch.autograd import Variable
+from datetime import datetime
+from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence
+
+def ortho_weight(ndim):
```
Comment (daemon) [inline esim/model.py]: Does `torch.nn.init` have orthogonal initialization?

File: esim/model.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    Used by norm_weights(below), in which case, we
+    are ensuring that the rows are orthogonal
+    (i.e W = U \Sigma V, U has the same
+    # of rows, V has the same # of cols)
+    """
+    W = np.random.randn(ndim, ndim)
+    u, s, v = np.linalg.svd(W)
+    return u.astype('float32')
+
+def norm_weight(nin, nout=None, scale=0.01, ortho=True):
```
Comment (daemon) [inline esim/model.py]: torch has this, right?

File: esim/model.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    """
+    if nout is None:
+        nout = nin
+    if nout == nin and ortho:
+        W = ortho_weight(nin)
+    else:
+        W = scale * np.random.randn(nin, nout)
+    return W.astype('float32')
+
+class LSTM_Cell(nn.Module):
```
Comment (daemon) [inline esim/model.py]: Why not `nn.LSTMCell`?

File: esim/model.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    def forward(self, input, h, c):
+        u = F.tanh(self.cx(input) + self.ch(h))
+        i = F.sigmoid(self.ix(input) + self.ih(h))
+        f = F.sigmoid(self.fx(input) + self.fh(h))
+        c = i*u + f*c
+        o = F.sigmoid(self.ox(input) + self.oh(h))
+        h = o * F.tanh(c)
+        return c, h
+
+class LSTM(nn.Module):
```
Comment (daemon) [inline esim/model.py]: Isn't this the same as `nn.LSTM`?

File: esim/model.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+        h = Variable(torch.zeros(x.size(1), x.size(2)))
+        c = Variable(torch.zeros(x.size(1), x.size(2)))
+        if torch.cuda.is_available():
+            h=h.to(self.device)
+            c=c.to(self.device)
+        all_hidden=[]
+        for step in range(x.size(0)):
+            input=x[step] # #sample x dim_emb
+            step_c, step_h=self.TreeCell(input, h, c)
+            h=x_mask[step][:,None] * step_h + (1. - x_mask[step])[:,None] * h
```
Comment (daemon) [inline esim/model.py]: Why not use the faster fused `nn.LSTM` with `pack_padded_sequence`?

File: esim/util.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+import torch
+import math
+import logging
+import numpy as np
+from os.path import expanduser
+from numpy import linalg as LA
+
+#tokenizer = nltk.tokenize.TreebankWordTokenizer()
+
+def pearson(x,y):
```
Comment (daemon) [inline esim/util.py]: SciPy has a Pearson's r routine, right?

File: esim/util.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+	sortedindex = sorted(range(len(probs)), key=probs.__getitem__)
+	sortedindex.reverse()
+
+	truepos=0
+	falsepos=0
+	for sortedi in sortedindex:
+		if test_data_label[sortedi]==True:
+			truepos+=1
+		elif test_data_label[sortedi]==False:
+			falsepos+=1
```
Comment (daemon) [inline esim/util.py]: This file is the only one that uses tabs; should use 4 spaces.

File: esim/util.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+	#data=(lsents,rsents,labels)
+	if not len(lsents)==len(rsents)==len(labels):
+		print('error!')
+		sys.exit()
+	clean_data = []
+	for i in range(len(lsents)):
+		clean_data.append((lsents[i], rsents[i], labels[i]))
+	return clean_data
+
+def readQuoradata(dir):
```
Comment (daemon) [inline esim/util.py]: Style is `read_quora_data`. Should we be putting _all_ the dataset utilities in a single `util.py`?

**Comment by daemon** [comment]
Mostly nitpicks. What's the reason for reimplementing an LSTM?

File: esim/util.py (hunk 9 above)
Comment (Victor0118) [reply esim/util.py]: Oh. Will try that. Thanks!

File: esim/util.py (hunk 11 above)
Comment (Victor0118) [reply esim/util.py]: Right. I should delete it since I never use read*() function.

**Comment by Victor0118** [comment]
@daemon Sorry. Just saw it. I think I just follow the raw implementation: https://github.com/lanwuwei/SPM_toolkit/tree/master/ESIM. No idea why the author reimplemented it.
